src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs

```diff
@@ -0,0 +1,87 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.\r
+// SPDX-License-Identifier: Apache-2.0
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using AWS.Deploy.Common;
+using AWS.Deploy.Common.IO;
+
+namespace AWS.Deploy.Orchestration.Utilities
+{
+    public interface ICloudApplicationNameGenerator
+    {
+        /// <summary>
+        /// Generates a valid candidate for <see cref="CloudApplication.Name"/> based on <paramref name="target"/>.
+        /// Name  is checked to ensure it is unique, ie <paramref name="existingApplications"/> doesn't already have it.
+        /// </summary>
+        /// <exception cref="ArgumentException">
+        /// Thrown if can't generate a valid name from <paramref name="target"/>.
+        /// </exception>
+        string GenerateValidName(ProjectDefinition target, List<CloudApplication> existingApplications);
+
+        /// <summary>
+        /// https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-using-console-create-stack-parameters.html
+        /// </summary>
+        bool IsValidName(string name);
+    }
+
+    public class CloudApplicationNameGenerator : ICloudApplicationNameGenerator
+    {
+        private readonly IFileManager _fileManager;
+        private readonly IDirectoryManager _directoryManager;
+        /// <summary>
+        /// https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-using-console-create-stack-parameters.html 
+        /// </summary>
+        private readonly Regex _validatorRegex = new ("^[a-zA-Z][a-zA-Z0-9-]{2,126}$", RegexOptions.Compiled);
+
+        public CloudApplicationNameGenerator(IFileManager fileManager, IDirectoryManager directoryManager)
+        {
+            _fileManager = fileManager;
+            _directoryManager = directoryManager;
+        }
+
+        public string GenerateValidName(ProjectDefinition target, List<CloudApplication> existingApplications)
```

Comment:
There is no check on the length of the application name.
```
The stack name is an identifier that helps you find a particular stack from a list of stacks. A stack name can contain only alphanumeric characters (case-sensitive) and hyphens. It must start with an alphabetic character and can't be longer than 128 characters.
```

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-using-console-create-stack-parameters.html

I think the generated name should match `^[a-zA-Z][-a-zA-Z0-9]{1,128}$`, I'm not saying to use regex.

Reply:
We can do the same validation on customer entered application for early feedback than trying to deploy and fail.

Reply:
1.  Length: `_validatorRegex` does check length: `{2,126}`  I was a little more restrictive than the regex you mentioned which uses `{1,128}`.
2. Early feedback.  Feedback and validation is done as the user enters/selects an Application Name.  They can't even see recommendations until a valid application name is selected.  See the screenshot in the PR Description.

Reply:
Initial regex was a bit too conservative.  Expanded to allow Names are short as 1 character and as long as 128